src/BindPattern.jl

```diff
@@ -193,25 +207,24 @@ function bind_pattern!(
         pattern = BoundTruePattern(location, source)
 
     elseif !(source isa Expr || source isa Symbol)
-        # a constant
-        bound_expression = BoundExpression(location, source)
-        pattern = BoundEqualValueTestPattern(input, bound_expression)
-
-    elseif is_expr(source, :macrocall) ||
-        is_expr(source, :quote) ||
-        (is_expr(source, :call) && source.args[1] == :Symbol) # the type `Symbol`
-        # Objects of Julia expression tree types.  We treat them as constants, but
-        # we should really match on their structure.  We might treat regular expressions
-        # specially, handle interpolation inside, etc.
-        # See #6, #30, #31, #32
-        bound_expression = BoundExpression(location, source)
-        pattern = BoundEqualValueTestPattern(input, bound_expression);
+        # a constant, e.g. a regular expression, version number, raw string, etc.
+        pattern = BoundIsMatchTestPattern(input, BoundExpression(location, source), false)
+
+    elseif is_expr(source, :macrocall)
+        # We permit custom string macros as long as they do not contain any unbound
+        # variables.  We accomplish that simply by expanding the macro.  Macros that
+        # interpolate, like lazy"", will fail because they produce a `call` rather
+        # than an object.  Also, we permit users to define macros that expand to patterns.
+        while is_expr(source, :macrocall)
+            source = macroexpand(binder.mod, source; recursive = false)
+        end
```

Comment:
is the reason we don't use `recursive = true` as we want the innermost expansion? does ignoring middle layers cause any potential issues?

Reply:
There are a couple of macros that can occur in nested contexts that we handle ourselves - specifically `@match_fail` and `@match_return`.  I don't know if this is really an issue, but this circumvents any possible issue.